### PR TITLE
`z_std_dma.c`

### PR DESCRIPTION
**Pull request by AngheloAlf**
I renamed `gCartHandle` to `carthandle` and `gDmaDataTable` to `dma_rom_ad`. Those names come from debug strings from OoT.

I stole a lot of docs from OoT

### CODE REVIEW

File: src/boot/z_std_dma.c [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+        }
+    } else {
+        // "Corresponding data does not exist"
+        DmaMgr_Error(req, NULL, "DATA DON'T EXIST", "該当するデータが存在しません");
+    }
+}
+
+void DmaMgr_ThreadEntry(UNUSED void* arg) {
+    OSMesg msg;
+    DmaRequest* sp34;
```
Comment (hensldm) [inline src/boot/z_std_dma.c:265]:
```suggestion
    DmaRequest* req;
```

File: src/boot/z_std_dma.c [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+    ioMsg.devAddr = vrom;
+    ioMsg.dramAddr = vram;
+    ioMsg.size = size;
+    B_800406B0_jp += size;
+
+    ret = osEPiStartDma(carthandle, &ioMsg, 0);
+    if (ret == 0) {
+        osRecvMesg(&queue, NULL, OS_MESG_BLOCK);
+        osInvalDCache(vram, size);
+    }
```
Comment (hensldm) [inline src/boot/z_std_dma.c:141]:
```suggestion
    if (ret != 0) {
        goto end;
    }
    
    osRecvMesg(&queue, NULL, OS_MESG_BLOCK);
    osInvalDCache(vram, size);
```
Perhaps this is me but this actually seems better so it mirrors the other usage.

File: src/boot/z_std_dma.c [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+
+    req->vrom = vrom;
+    req->vram = vram;
+    req->size = size;
+    req->unk_14 = 0;
+    req->mq = mq;
+    req->unk_1C = arg6;
+
+    if ((vram == NULL) || (osMemSize < ((uintptr_t)vram + size + 0x80000000)) || (vrom % 2 != 0) ||
```
Comment (hensldm) [inline src/boot/z_std_dma.c:309]:
```suggestion
    if ((vram == NULL) || (osMemSize < (OS_K0_TO_PHYSICAL(vram) + size)) || (vrom % 2 != 0) ||
```

File: src/boot/z_std_dma.c [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+/**
+ * Synchronous DMA Request with source file and line info for debugging.
+ *
+ * @see DmaMgr_RequestSync
+ */
+s32 DmaMgr_RequestSyncDebug(void* vram, RomOffset vrom, size_t size, const char* filename, s32 line) {
+    DmaRequest req;
+    s32 ret;
+    OSMesgQueue mq;
+    OSMesg sp30[1];
```
Comment (hensldm) [inline src/boot/z_std_dma.c:453]:
```suggestion
    OSMesg msg[1];
```

File: src/boot/z_std_dma.c [hunk 5]
```diff
@@ -0,0 +1,470 @@
...
+ *
+ * @return 0 if successful, -1 if the DMA could not be queued with the PI Manager.
+ */
+s32 DmaMgr_DmaRomToRam(RomOffset vrom, void* vram, size_t size) {
+    OSIoMesg ioMsg;
+    OSMesgQueue queue;
+    OSMesg msg[1];
+    s32 ret;
+    size_t buffSize = gDmaMgrDmaBuffSize;
+    DmaRequest sp3C;
```
Comment (hensldm) [inline src/boot/z_std_dma.c:86]:
```suggestion
    DmaRequest req;
```

File: include/z_std_dma.h [hunk 6]
```diff
@@ -12,26 +25,28 @@ typedef struct DmaEntry {
...
+s32 DmaMgr_GetOvlOffsets(RomOffset vromStart, RomOffset* vromEnd, RomOffset* ovlStart, RomOffset* ovlEnd);
+void DmaMgr_Init(void);
+void DmaMgr_Stop(void);
+void DmaMgr_RequestAsync(DmaRequest* req, void* vram, RomOffset vrom, size_t size, s32 arg4, OSMesgQueue* mq, OSMesg msg, const char* filename, s32 line);
+s32 DmaMgr_RequestSyncDebug(void* vram, RomOffset vrom, size_t size, const char* filename, s32 line);
+
+
+#define DMAMGR_DEFAULT_BUFSIZE ALIGN16(0x2000)
+
+// .data
```
Comment (hensldm) [inline include/z_std_dma.h:45]:
```suggestion
```
Probably don't need these?

File: src/boot/z_std_dma.c [hunk 7]
```diff
@@ -0,0 +1,470 @@
...
+    do {
+        DmaEntry* entry = dma_rom_ad;
+        s32 count = 0;
+
+        for (; entry->vromEnd != 0; entry++) {
+            count++;
+        }
+
+        sNumDmaEntries = count;
+    } while (0);
```
Comment (hensldm) [inline src/boot/z_std_dma.c:411]: Do you need the full `do while(0);`? Perhaps just scoped works (though not sure it is better). Should we have a `fake` label here?

**Review by hensldm — COMMENTED** [review]
Missed a few things on the first pass

File: src/boot/z_std_dma.c (hunk 7 above)
Comment (AngheloAlf) [reply src/boot/z_std_dma.c:411]: Doesn't need the `do while(0)` but it needs the scope. We usually don't do fake labels for scopes like this. Also, MM has this same scope https://github.com/zeldaret/mm/blob/cbf9d98dba61924625629061f7680e3508fc2701/src/boot_O2_g3/z_std_dma.c#L220-L230

**Review by hensldm — APPROVED** [review]
:)

**Pull request merged**